File: voice-note-client/android/app/src/main/kotlin/com/spark/suikouji/audio/NativeAudioPlugin.kt
package com.spark.suikouji.audio

import android.content.Context
import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

class NativeAudioPlugin : MethodCallHandler, EventChannel.StreamHandler {
    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var eventSink: EventChannel.EventSink? = null
    private val mainHandler = Handler(Looper.getMainLooper())

    private lateinit var runtimeController: AudioRuntimeController

    fun attach(context: Context, methodChannel: MethodChannel, eventChannel: EventChannel) {
        this.methodChannel = methodChannel
        this.eventChannel = eventChannel
        this.runtimeController = AudioRuntimeController(context) { event, payload ->
            val sink = eventSink ?: return@AudioRuntimeController
            val envelope = mutableMapOf<String, Any?>(
                "event" to event,
                "sessionId" to (payload["sessionId"] as? String ?: ""),
                "requestId" to payload["requestId"],
                "timestamp" to (payload["timestamp"] as? Long ?: System.currentTimeMillis()),
                "data" to (payload["data"] ?: emptyMap<String, Any?>()),
                "error" to payload["error"],
            )
            mainHandler.post { sink.success(envelope) }
        }
        this.methodChannel.setMethodCallHandler(this)
        this.eventChannel.setStreamHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        val args = (call.arguments as? Map<*, *>)?.entries
            ?.associate { it.key.toString() to it.value } ?: emptyMap<String, Any?>()

        val output = when (call.method) {
            "initializeSession" -> runtimeController.initializeSession(args)
            "disposeSession" -> runtimeController.disposeSession(args)
            "setAsrMuted" -> runtimeController.setAsrMuted(args)
            "playTts" -> runtimeController.playTts(args)
            "stopTts" -> runtimeController.stopTts(args)
            "setBargeInConfig" -> runtimeController.setBargeInConfig(args)
            "getDuplexStatus" -> runtimeController.getDuplexStatus()
            "switchInputMode" -> runtimeController.switchInputMode(args)
            "getLifecycleSnapshot" -> runtimeController.getLifecycleSnapshot()
            "restoreLifecycleSnapshot" -> runtimeController.restoreLifecycleSnapshot(args)
            "startAsrStream" -> runtimeController.startAsrStream(args)
            "commitAsr" -> runtimeController.commitAsr()
            "stopAsrStream" -> runtimeController.stopAsrStream()
            "startCapture" -> runtimeController.startCapture(args)
            "stopCapture" -> runtimeController.stopCapture(args)
            else -> null
        }

        if (output != null) {
            result.success(output)
        } else {
            result.notImplemented()
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }
}

File: voice-note-client/ios/Runner/Audio/NativeAudioPlugin.swift
import Flutter
import Foundation

final class NativeAudioPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
  private static let methodChannelName = "voice_note/native_audio"
  private static let eventChannelName = "voice_note/native_audio/events"

  private var eventSink: FlutterEventSink?
  private lazy var runtimeController = AudioRuntimeController { [weak self] event, payload in
    guard let self = self else { return }
    guard let sink = self.eventSink else { return }
    var envelope: [String: Any] = [
      "event": event,
      "sessionId": payload["sessionId"] as? String ?? "",
      "requestId": payload["requestId"] as Any,
      "timestamp": payload["timestamp"] as? Int ?? Int(Date().timeIntervalSince1970 * 1000),
      "data": payload["data"] as? [String: Any] ?? [:],
      "error": payload["error"] as Any
    ]
    // Keep envelope shape identical to Android.
    if envelope["requestId"] == nil { envelope["requestId"] = NSNull() }
    if envelope["error"] == nil { envelope["error"] = NSNull() }
    // Flutter requires platform channel messages on the platform (main) thread.
    if Thread.isMainThread {
      sink(envelope)
    } else {
      DispatchQueue.main.async { sink(envelope) }
    }
  }

  static func register(with registrar: FlutterPluginRegistrar) {
    let plugin = NativeAudioPlugin()
    let methodChannel = FlutterMethodChannel(
      name: methodChannelName,
      binaryMessenger: registrar.messenger()
    )
    let eventChannel = FlutterEventChannel(
      name: eventChannelName,
      binaryMessenger: registrar.messenger()
    )
    registrar.addMethodCallDelegate(plugin, channel: methodChannel)
    eventChannel.setStreamHandler(plugin)
  }

  func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    let args = call.arguments as? [String: Any] ?? [:]
    let output: [String: Any]?
    switch call.method {
    case "initializeSession":
      output = runtimeController.initializeSession(args: args)
    case "disposeSession":
      output = runtimeController.disposeSession(args: args)
    case "setAsrMuted":
      output = runtimeController.setAsrMuted(args: args)
    case "playTts":
      output = runtimeController.playTts(args: args)
    case "stopTts":
      output = runtimeController.stopTts(args: args)
    case "setBargeInConfig":
      output = runtimeController.setBargeInConfig(args: args)
    case "getDuplexStatus":
      output = runtimeController.getDuplexStatus()
    case "switchInputMode":
      output = runtimeController.switchInputMode(args: args)
    case "getLifecycleSnapshot":
      output = runtimeController.getLifecycleSnapshot()
    case "restoreLifecycleSnapshot":
      output = runtimeController.restoreLifecycleSnapshot(args: args)
    case "startAsrStream":
      output = runtimeController.startAsrStream(args: args)
    case "commitAsr":
      output = runtimeController.commitAsr()
    case "stopAsrStream":
      output = runtimeController.stopAsrStream()
    case "startCapture":
      output = runtimeController.startCapture(args: args)
    case "stopCapture":
      output = runtimeController.stopCapture(args: args)
    default:
      output = nil
    }

    if let output = output {
      result(output)
    } else {
      result(FlutterMethodNotImplemented)
    }
  }

  func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    eventSink = events
    return nil
  }

  func onCancel(withArguments arguments: Any?) -> FlutterError? {
    eventSink = nil
    return nil
  }
}
